scsi rebase

Ablu opened the pull request:
### Summary of the PR

This is a rebase of https://github.com/rust-vmm/vhost-device/pull/4.

### Requirements

Before submitting your PR, please make sure you addressed the following
requirements:

- [x] All commits in this PR are signed (with `git commit -s`), and the commit
  message has max 60 characters for the summary and max 75 characters for each
  description line.
- [x] All added/changed functionality has a corresponding unit/integration
  test.
- [x] All added/changed public-facing functionality has entries in the "Upcoming 
  Release" section of CHANGELOG.md (if no such section exists, please create one).
- [x] Any newly added `unsafe` code is properly documented.

Ablu commented:
> * All commits in this PR are signed (with `git commit -s`), and the commit
>   message has max 60 characters for the summary and max 75 characters for each
>   description line.

Is there documentation of what signing means? I assume DCO, but I find it a bit weird to sign things without that having specified somewhere :)

Ablu commented on crates/scsi/rustfmt.toml:
```diff
@@ -0,0 +1,4 @@
+imports_granularity = "Crate"
+group_imports = "StdExternalCrate"
+wrap_comments = true
+normalize_comments = true
```
Do we need/want this?

stsquad replied:
I think all the clippy and rustfmt rules are covered by the common CI stuff. If that is the case I don't think we need to make it explicit here as the rust-vmm coding style is documented for all crates.

Ablu commented on crates/scsi/src/main.rs:
```diff
@@ -0,0 +1,400 @@
+#![warn(missing_debug_implementations)]
+#![allow(clippy::cast_possible_truncation)]
+#![allow(clippy::module_name_repetitions)]
+#![allow(clippy::non_ascii_literal)]
```
TODO: review

Ablu commented on crates/scsi/src/main.rs:
```diff
@@ -0,0 +1,400 @@
+impl VhostUserBackendMut<VringRwLock> for VhostUserScsiBackend {
+    fn num_queues(&self) -> usize {
+        let num_request_queues = 1;
+        2 + num_request_queues
```
Is a fixed queue number fine for a first version? Or does this need additional work?

stsquad replied:
I think as long as we document the limitation along with a working QEMU command line in the README we can go with it for now. A comment wouldn't go amiss though.

Ablu replied:
I already adjusted the QEMU commandline to accord for it. Will add another comment here and mention it in the documentation more clearly.

Ablu commented on crates/scsi/src/main.rs:
```diff
@@ -0,0 +1,400 @@
+    }
+
+    fn protocol_features(&self) -> VhostUserProtocolFeatures {
+        VhostUserProtocolFeatures::MQ
```
Do we need more here in a first version?

Ablu commented on crates/scsi/src/main.rs:
```diff
@@ -0,0 +1,400 @@
+    fn features(&self) -> u64 {
+        // TODO: Any other ones worth implementing? EVENT_IDX and INDIRECT_DESC
+        // are supported by virtiofsd
+        1 << VIRTIO_F_VERSION_1 | VhostUserVirtioFeatures::PROTOCOL_FEATURES.bits() | 1 << 2
```
https://github.com/rust-vmm/vhost-device/pull/4#discussion_r763929627
```
Yes you probably want the full set:

    1 << VIRTIO_F_VERSION_1
        | 1 << VIRTIO_F_NOTIFY_ON_EMPTY
        | 1 << VIRTIO_RING_F_INDIRECT_DESC
        | 1 << VIRTIO_RING_F_EVENT_IDX
        | VhostUserVirtioFeatures::PROTOCOL_FEATURES.bits()
drop the hardcoded 1 << 2```

stsquad replied:
Yep definitely use the constants.

Ablu replied:
Requires new bindings: https://github.com/rust-vmm/vm-virtio/pull/231

Ablu commented on crates/scsi/src/main.rs:
```diff
@@ -0,0 +1,400 @@
+
+    fn set_event_idx(&mut self, enabled: bool) {
+        // Should always be true until we support EVENT_IDX in features.
+        assert!(!enabled);
```
TODO: turn into assert!(cond, error)

Ablu commented on crates/scsi/src/main.rs:
```diff
@@ -0,0 +1,400 @@
+        assert!(thread_id == 0);
+
+        // unwrap: only fails if the lock is poisoned, in which case we already panicked
+        // somewhere else
```
stale comment?

Ablu commented on crates/scsi/src/main.rs:
```diff
@@ -0,0 +1,400 @@
+            .signal_used_queue()
+            .map_err(|e| io::Error::new(ErrorKind::Other, e))?;
+
+        Ok(false) // TODO: what's this bool? no idea. virtiofd-rs returns false
```
review

Ablu replied:
clarification happens at: https://github.com/rust-vmm/vhost/issues/144. I guess this is fine and I can drop the comment.

Ablu commented on crates/scsi/src/scsi/emulation/block_device.rs:
```diff
@@ -0,0 +1,547 @@
+impl BlockDevice {
+    pub const fn new(file: File) -> Self {
+        // TODO: trying 4096 logical/physical for now. May need to fall
+        // back to 512 logical/4096 physical for back compat.
```
This looks stale, given that 512 is passed below? :thinking:

Ablu commented on crates/scsi/src/scsi/emulation/block_device.rs:
```diff
@@ -0,0 +1,547 @@
+                    ReportSupportedOpCodesMode::All => {
+                        let cmd_len = if rctd { 20 } else { 8 };
+                        // unwrap is safe: we're never going to have (2^32 / 20) ~= 2^27 opcodes
+                        let len = u32::try_from(OPCODES.len() * cmd_len).unwrap();
```
in general, many of the unwrap() comments could probably turned into a .expect()

Ablu reviewed (commented):
left some comments for myself

stsquad commented on crates/scsi/src/virtio.rs:
```diff
@@ -0,0 +1,267 @@
+#[derive(Debug)]
+pub enum VirtioScsiError {
+    // VirtioError(VirtioError),
+    // ScsiError(ScsiError)
```
We certainly need to have a reasonable error set for reporting.

Ablu replied:
Right now this file does not need it though. It only implements some Read/Write traits that go through `io::Result`. So I would drop this for now. It was just wrapping errors anyway and proper error reporting likely needs to be considered on the grand scheme of the PR

stsquad replied:
seems fair

Ablu commented:
Rough changelog:
* squashed things, splitted into (hopefully sensible) chunks
* moved tests under emulation (thats what they were testing)
* implemented idx
* fixed most previous comments

TODOs:
- [ ] fix author ship... Somewhere during rebase I lost the authorship
- [ ] review `pub`s. How do we want this crate to be consumed? @stsquad: Do we want this to just be a daemon or shall the interfaces be exported as lib (assumed the latter for now)?
- [ ] throughout self-review

Ablu commented:
@Gaelan: Did I get the authorship right? Do we need to credit other people/projects from which things were adopted?

Ablu commented on crates/scsi/src/scsi/emulation/block_device.rs:
```diff
@@ -0,0 +1,548 @@
+}
+
+impl BlockDevice {
+    pub const fn new(file: File) -> Self {
```
To clarify: Is `File` the correct abstraction level here? Or should we use some `Read`-like trait as backend here? That could allow for some more options: testing with super large (fake) backend, network backed implementations (not sure how great the idea is?).

Not exactly sure what level of synchronization is required for write support or a potential multithreading addition...

Gaelan commented:
Many thanks for picking this back up!

> Did I get the authorship right? Do we need to credit other people/projects from which things were adopted?

All the code here was written by me (as best I recall; I'm generally pretty good about noting if it isn't). The work was done as part of GSoC, but I don't believe any copyright assignment takes place there.

Incidentally, I had a `scsi-write` branch that may be of interest; there was sufficient write support to boot and run Debian, though I can't vouch for its correctness in the face of writeback failures - IIRC there's a tricky mismatch here between the level of detail SCSI expects and what the POSIX file APIs give us. I think I might just `fsync` everything, which is probably slow-but-good-enough?

Gaelan commented:
Please let me know if you have any questions, by the way! This was a couple years ago now, so no promises, but happy to try and remember what I was thinking.

Ablu commented:
> All the code here was written by me (as best I recall; I'm generally pretty good about noting if it isn't). The work was done as part of GSoC, but I don't believe any copyright assignment takes place there.

Yeah, you had sign-offs on the relevant commits, mostly tried to make sure that the way I split up stuff is ok with you (since I put your name as main author). That said, we might be missing SPDX-License-Identifiers (at least current code has it, do not find anything in https://github.com/rust-vmm/community/blob/main/CONTRIBUTING.md). I would add: `Apache-2.0 or BSD-3-Clause` to all files, ok?


> Incidentally, I had a `scsi-write` branch that may be of interest; there was sufficient write support to boot and run Debian, though I can't vouch for its correctness in the face of writeback failures - IIRC there's a tricky mismatch here between the level of detail SCSI expects and what the POSIX file APIs give us. I think I might just `fsync` everything, which is probably slow-but-good-enough?

I guess for now, I would aim at getting the readonly state merged. So I have not spent any thoughts on writes. Will check it out when I do so!



> Please let me know if you have any questions, by the way! This was a couple years ago now, so no promises, but happy to try and remember what I was thinking.

Sure. Right now I focused on getting stuff work properly. Will let you know when I stumble over something. Feel free to provide review on the stuff that I changed!

Ablu commented on crates/scsi/src/scsi/emulation/mod.rs:
```diff
@@ -0,0 +1,75 @@
+            .get(lun as usize)
+            .map_or(&MissingLun, |x| x.as_ref());
+
+        lun.execute_command(req, self)
```
This is a bit weird and breaks once one attempts to make `self` mutable... `self` is only needed for `Command::ReportLuns`... Which does not seem to be very lun-specific (saying that without knowing too much about SCSI)? So probably at least this command should be handled on this abstraction level...

Ablu commented:
Changelog (in order of significance):
- Split the emulation interface into handling LUN-independent and LUN-specific messages
  - This allows the LUN-specific part to be oblivious to the target
  - Allowed for some de-duplication too
- Create own trait to decouple from `io::File`
- De-boilerplated the daemon by moving parsing to virtio.rs
  - Still requires some boilerplate to copy between virtio and scsi datastructures
  - Could be further simplified if we are fine with scsi-virtio-specific stuff returning datastructures of src/scsi/ (currently they were kept separated on purpose)
- Removed no longer needed suppressing of linting (either due to simplified code or newer linter version)
- Random cleanups here  and there

Ablu commented:
CI seems to be broken by dirty permissions from left-over jobs?

Ablu commented:
Open clarification items:

- [ ] SPDX identifiers?
- [ ] abstraction-levels: What should be public? Which parts do we want to allow other people to use?
  - "Just the daemon?" vs "daemon, virtio-scsi abstraction, target interface, inner-emulation interfaces (`LogicalUnit`)"?
- [ ] How to test the daemon?
- [ ] Meaning of signoff is not documented.
- [ ] No CHANGELOG available.

Gaelan commented on crates/scsi/src/scsi/emulation/command.rs:
```diff
@@ -0,0 +1,579 @@
+#[derive(Debug)]
+pub enum Command {
+    LunIndependentCommand(LunIndependentCommand),
+    LunSpecificCommand(LunSpecificCommand),
```
It's worth noting that several of the "LUN-specific" commands here are used for multiple LUN types, not just block devices. (TEST UNIT READY, REQUEST SENSE, REPORT SUPPORTED OPERATION CODES, and INQUIRY are required for all LUN types; commands like READ CAPACITY and READ are used by several storage-like devices.)

So if we ever wanted to do more than just block devices, we'd need to do some more thinking about how we want to handle that. Unfortunately we can't just parse a command without knowing the intended device type, as sometimes operation codes are reused.

Anyway, cross that bridge when we get to it - just potentially useful background.

Gaelan commented on crates/scsi/src/scsi/emulation/tests/generic.rs:
```diff
@@ -0,0 +1,105 @@
+        },
+    );
+
+    if let CmdError::DataIn(e) = res.unwrap_err() {
```
Oops, leftover comment from an earlier review i didn't mean to submit here, please disregard

Ablu commented:
> abstraction-levels: What should be public? Which parts do we want to allow other people to use?
> * "Just the daemon?" vs "daemon, virtio-scsi abstraction, target interface, inner-emulation interfaces (`LogicalUnit`)"?

Currently, we have multiple layers of abstraction in here:
* the daemon that allows connection to the socket, parses things using the `virtio` module and forwards things to the `EmulatedTarget` after registering a `BlockDevice` for each image.
* `virtio` that abstracts some of the virtio-scsi related parsing and provides some convenience features
* `scsi::Target` that abstracts over the actual backend (only consists of `execute_command`)
* `scsi::emulation::target::EmulationTarget` that implements `scsi::Target` by maintaining a list of `LogicalUnit`s, handling `REPORT_LUNS` and forwarding other commands to the `LogicalUnit`s.
* `scsi::emulation::block_device::BlockDevice` that emulates an actual drive against a `BlockDeviceBackend`
* `scsi::emulation::block_device::FileBackend` that implements `BlockDeviceBackend`

Comparing that against https://github.com/rust-vmm/vm-virtio/tree/main/crates/devices/virtio-blk/src, my feeling is that there the `virtio` and high-level device-specific abstractions are sharing the same data structures.

With that, there are a series of choices that could be made:
1. About virtio parsing + `scsi::Target`:
   * Move it to https://github.com/rust-vmm/vm-virtio/tree/main/crates/devices.
      * Will probably require some extensive review and additions for features that we do not cover.
      * Unifying the two interface, would make the main.rs more simple, since no conversion is necessary.
   * Keep it here and public.
      * People could depend on this without consuming the rest, but it would not be as clean as pulling it from a separate crate...
      * We would probably at least still need to document this a bit better...
    * Keep it here, but mark it private.
      * Only our daemon would profit from it then.
      * we could mature it over time and eventually move it to https://github.com/rust-vmm/vm-virtio/tree/main/crates/devices.
    * No real benefit from defining separate `scsi` (virtio-less) and `scsi-virtio` interfaces though...
2. About the emulation-specific abstraction
    * Exposing these traits probably only makes sense if we also provide enough other support for the virtio layers in order to allow people to write their own daemon.
    * If we do not do that, we should probably just mark that API as crate-private.

I got no strong feelings about any of these options and are open to opinions.

Ablu commented:
> I would add: `Apache-2.0 or BSD-3-Clause` to all files, ok?

@Gaelan: Can you confirm that your code is of that license (not finding where the implications of the sign-off are stated... Probably this is already agreed to by the sign-off, but better to double check...)?
I would then add a:
```
// SPDX-License-Identifier: Apache-2.0 or BSD-3-Clause
```

tag to all files.

Gaelan commented:
Yes.

Ablu commented on crates/scsi/src/virtio.rs:
```diff
@@ -0,0 +1,307 @@
+}
+
+/// A `Write` implementation that writes to the memory indicated by a virtio
+/// descriptor chain.
```
reference https://github.com/rust-vmm/vm-virtio/pull/33

stsquad commented:
OK first a couple of minor things:

 - do we want to rename this to vhost-device-scsi to be more consistent with the others (maybe not as vhost-user-vsock also exists)
 - can we agree on standard parameter conventions, e.g. --socket-path even if we don't need to handle multiple

Finally when running with:

```
./qemu-system-aarch64 -cpu max,pauth-impdef=on -machine type=virt,virtualization=on,gic-version=3 -display none -serial mon:stdio -netdev user,id=unet,hostfwd=tcp::2222-:
22,hostfwd=tcp::1234-:1234 -device virtio-net-pci,netdev=unet -kernel ~/lsrc/linux.git/builds/arm64.initramfs/arch/arm64/boot/Image -append "console=ttyAMA0 init=/sbin/init"
 -m 4096 -object memory-backend-file,id=mem,size=4G,mem-path=/dev/shm,share=on -numa node,memdev=mem -smp 1 -chardev socket,id=scsi,path=/tmp/scsi.sock -device vhost-user-sc
si-pci,chardev=scsi,num_queues=1,param_change=off -d trace:vhost_\*,trace:vhost_user\* 
```

I see the following crash:

```
✗  env RUST_LOG=trace RUST_BACKTRACE=1 ./target/debug/vhost-user-scsi -r /tmp/scsi.sock test.img
[2023-03-31T12:46:53Z INFO  vhost_user_scsi] Memory updated - guest probably booting
[2023-03-31T12:46:53Z ERROR vhost_user_scsi] Ignoring descriptor on queue 1
thread 'vring_worker' panicked at 'should fit into cdb: TryFromSliceError(())', /home/alex/lsrc/virtio.rust/vhost-device.git/crates/scsi/src/virtio.rs:103:18
stack backtrace:
   0: rust_begin_unwind
             at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/std/src/panicking.rs:575:5
   1: core::panicking::panic_fmt
             at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/core/src/panicking.rs:64:14
   2: core::result::unwrap_failed
             at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/core/src/result.rs:1791:5
   3: core::result::Result<T,E>::expect
             at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/core/src/result.rs:1070:23
   4: vhost_user_scsi::virtio::Request::parse
             at ./crates/scsi/src/virtio.rs:101:18
   5: vhost_user_scsi::VhostUserScsiBackend::handle_request_queue
             at ./crates/scsi/src/main.rs:117:30
   6: vhost_user_scsi::VhostUserScsiBackend::process_request_queue
             at ./crates/scsi/src/main.rs:92:13
   7: <vhost_user_scsi::VhostUserScsiBackend as vhost_user_backend::backend::VhostUserBackendMut<vhost_user_backend::vring::VringRwLock>>::handle_event
             at ./crates/scsi/src/main.rs:270:25
   8: <std::sync::rwlock::RwLock<T> as vhost_user_backend::backend::VhostUserBackend<V,B>>::handle_event
             at /home/alex/.cargo/registry/src/github.com-1ecc6299db9ec823/vhost-user-backend-0.8.0/src/backend.rs:393:9
   9: <alloc::sync::Arc<T> as vhost_user_backend::backend::VhostUserBackend<V,B>>::handle_event
             at /home/alex/.cargo/registry/src/github.com-1ecc6299db9ec823/vhost-user-backend-0.8.0/src/backend.rs:262:9
  10: vhost_user_backend::event_loop::VringEpollHandler<S,V,B>::handle_event
             at /home/alex/.cargo/registry/src/github.com-1ecc6299db9ec823/vhost-user-backend-0.8.0/src/event_loop.rs:223:9
  11: vhost_user_backend::event_loop::VringEpollHandler<S,V,B>::run
             at /home/alex/.cargo/registry/src/github.com-1ecc6299db9ec823/vhost-user-backend-0.8.0/src/event_loop.rs:197:20
  12: vhost_user_backend::handler::VhostUserHandler<S,V,B>::new::{{closure}}
             at /home/alex/.cargo/registry/src/github.com-1ecc6299db9ec823/vhost-user-backend-0.8.0/src/handler.rs:132:32
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

Ablu commented:
> I see the following crash:

TODO:
[ ] support `CONFIG` in order to prevent hardcoding CDB_SIZE

Ablu commented:
> [ ] support `CONFIG` in order to prevent hardcoding CDB_SIZE

turned out to be a different issues, dropping this as TODO for first milestone. Linux and QEMU currently seem to both use a fixed size


# Changelog:
- Added license tags
- Fixed size of CDB (@stsquad this should fix your crash)
- Started writing tests (more are needed)
- Refactored daemon to match code structure of other daemons
- Renamed to vhost-device-scsi
- Renamed sock -> socket-path and turned into a named argument

# TODOs:
- [ ] make API private
- [ ] add changelog file
- [ ] expand tests

Gaelan commented on crates/scsi/src/virtio.rs:
```diff
@@ -0,0 +1,358 @@
+        } else if bytes[0] == 0x1 {
+            let target = bytes[1];
+            // bytes[2..3] is a normal SCSI single-level lun
+            if (bytes[2] & 0b0100_0000) == 0 {
```
Ok, retracing my steps here: The format of the 8 bytes passed in here is defined in the [virtio-scsi spec](https://docs.oasis-open.org/virtio/virtio/v1.1/csprd01/virtio-v1.1-csprd01.html#x1-308001r311), which says:

> lun addresses the REPORT LUNS well-known logical unit, or a target and logical unit in the virtio-scsi device’s SCSI domain. When used to address the REPORT LUNS logical unit, lun is 0xC1, 0x01 and six zero bytes. The virtio-scsi device SHOULD implement the REPORT LUNS well-known logical unit.
> When used to address a target and logical unit, the only supported format for lun is: first byte set to 1, second byte set to target, third and fourth byte representing a single level LUN structure, followed by four zero bytes. With this representation, a virtio-scsi device can serve up to 256 targets and 16384 LUNs per target. The device MAY also support having a well-known logical units in the third and fourth byte.

The encoding for `bytes[2..3]` appears to match the "Single level LUN structure using flat space addressing method", found in Table 15 of SAM-6 revision 7 (the draft I was developing against) - it's just `01b` followed by a 14-bit LUN number.

Gaelan replied:
Oh huh, this code would silently accept `11b` as well - I very much suspect that was a mistake.

Ablu replied:
Ok. Things are making a bit more sense after lunch... I got confused by both structures being 8 bytes, but for virtio it looks like the `0x1, <target>` prefix has little else to do with the LUN in the scsi spec, then the next two bytes are the _actual_ LUN. Will add a check to make sure that we only support `0b01` right now.

Gaelan replied:
Incidentally, I have no idea what the "REPORT LUNS well-known LUN" as described in virtio-scsi is meant to do - it doesn't have any place to specify a target, so presumably the intent is that it returns LUNs from all targets? But presumably it would return LUNs in the SCSI format, which doesn't include a target number (because SCSI assumes the command is being sent to a LUN on a single target). In any case, I'm not sure anybody uses it.

Ablu commented:
# Changelog
- test skeleton for testing vhu_scsi.rs
- fixed clippy warnings
- stricter LUN format check

# TODO
- [ ] some error-case tests

Ablu commented:
# Changelog:
- Extended tests with some error conditions

I think this could now be in a mergable state (pending any review comments of course)

Ablu commented:
# Changelog:
- Added support for CONFIG for testing with @stsquad's vhost-user-generic series (and for later Xen support)

Ablu commented:
# Changelog:

- rebased on origin/main
- removed some fields that are not actually used by the current implementation
- added WRITE(10), WRITE SAME(16) and SYNCHRONIZE CACHE(10) support
- fixed an issue found by fuzzing (see https://github.com/Ablu/vhost-device/tree/scsi-fuzzing for the quick&dirty fuzzing code)

Ablu commented:
# Changelog

- fixed build
- added some testing infrastructure

vireshk commented on crates/scsi/Cargo.lock:
```diff
@@ -0,0 +1,458 @@
+# This file is automatically @generated by Cargo.
```
There shouldn't be a crate specific Cargo.lock file, isn't it ?

vireshk commented on crates/scsi/Cargo.toml:
```diff
@@ -0,0 +1,28 @@
+name = "vhost-device-scsi"
+version = "0.1.0"
+authors = ["Gaelan Steele <gbs@canishe.com>"]
+edition = "2018"
```
Other crates have 2021 here.

Can you please cross check with any other crate, like gpio and see if there are any differences that we can get rid of ?

vireshk commented on crates/scsi/Cargo.toml:
```diff
@@ -0,0 +1,28 @@
+tempfile = "3.2.0"
+virtio-queue = { version = "0.7", features = ["test-utils"] }
+
+
```
Can we remove blank lines at the end of the file ?

vireshk commented on crates/scsi/README.md:
````diff
@@ -0,0 +1,48 @@
+Run the vhost-user-scsi daemon:
+
+```
+vhost-user-scsi -r /tmp/vhost-user-scsi.sock /path/to/image.raw /path/to/second-image.raw ...
````
don't need --socket-path ?

vireshk commented on crates/scsi/src/backend.rs:
```diff
@@ -0,0 +1,159 @@
+    sync::{Arc, RwLock},
+};
+
+use crate::scsi::emulation::{
```
We are following this ordering of using modules in this crate:
- std
- other dependencies
- super/crate ones.

Move this one along with crate::vhu_scsi please.

vireshk commented on crates/scsi/src/backend.rs:
```diff
@@ -0,0 +1,159 @@
+    let mut backend = VhostUserScsiBackend::new();
+    let mut target = EmulatedTarget::new();
+    if args.images.len() > 256 {
+        error!("");
```
Don't want to print anything here ?

vireshk commented on crates/scsi/src/backend.rs:
```diff
@@ -0,0 +1,159 @@
+        // encoding formats.
+        return Err(Error::TooManyLUNs);
+    }
+    if !args.read_only {
```
I would add blank lines after every block of code, like the if block above and this one.

vireshk commented on crates/scsi/src/main.rs:
```diff
@@ -0,0 +1,10 @@
+mod virtio;
+
+fn main() -> backend::Result<()> {
+    backend::scsi_init()
```
I don't see much value in creating backend.rs, maybe just merge main.rs and backend.rs.

I had to do it for gpio crate to get rid of some Musl specific configurations.

vireshk commented on crates/scsi/test/.gitignore:
```diff
@@ -0,0 +1,2 @@
+results/
```
Why is this required ? Also .containerignore ?

Ablu reviewed (commented):
Thanks @vireshk.

Fixed/reacted to all comments.

While rebasing I noticed missing tests. While adding them I went into a rabbit hole of trying to fix some API weirdness around the `<W, R>` generics. I ended up dropping them in favor of dynamic dispatch. That simplifies the code quite a bit. And I think spending too much time thinking through the monomorphization without any benchmarks in place is not worth scarifying code clarity.

# Changelog
- Dropped scsi-specific .lock file
- Match the gpiod Cargo.toml fields
- Fixed invocation in docs
- Merged backend.rs into main.rs again
- Switched from generics over Read/Write to dynamic dispatch
- Dropped data_in/data_out from Request and turned them into ordinary function arguments
- Minor API cleanup (some `pub`/`pub(crate)`, `&mut` changes)
- Made newlines more consistent
- Dropped (for now) unused `unmap` field
- Removed some left-over comment
- Added tests for Write(10), WriteSame(16)
- Dropped `test_all*` tests that would require annoying updates every time a new command is added
- Reordered usings (`cargo fmt -- --config group_imports=StdExternalCrate`)

epilys commented on crates/scsi/src/scsi/emulation/command.rs:
```diff
@@ -0,0 +1,681 @@
+    UnitSerialNumber,                // *
+    ZonedBlockDeviceCharacteristics, // *
+}
+// starred ones are ones Linux will use if availible
```
```suggestion
// starred ones are ones Linux will use if available
```

Ablu reviewed (commented):
@epilys: Thanks!

# Changelog
- Introduced new-types for ByteOffset, BlockOffset, BlockSize

Ablu commented:
CI reports dropped coveraged. But that seems to happen because the CI reports slightly different percentages compared to what my local toolchain reports. I will adjust the number to make CI happy once this is reviewed. Overall the coverage improved, the failure is only happening because I bumped the percentage to what the script tells me locally.

Ablu commented:
# Changelog
- Rebased on top of main
- Bumped dependencies to match other crates
- Dropped unneeded dependency on queue test-utils

Ablu commented:
# Changelog
- Hopefully fixed coverage

vireshk commented:
This is just too much (and complex) code to review. I have looked at it from the perspective of vhost-device stuff and that it is implemented in a way similar to other crates in the workspace.

Approving it now.

@stsquad @mathieupoirier @stefano-garzarella

stsquad commented:
> This is just too much (and complex) code to review. I have looked at it from the perspective of vhost-device stuff and that it is implemented in a way similar to other crates in the workspace.
> 
> Approving it now.
> 
> @stsquad @mathieupoirier @stefano-garzarella

I think its had some good cycles of review but there is a trade off to a growing stack of commits vs incremental updates. I'm happy with merging this so follow on enhancements aren't forever blocked.
